File: replaceWord.cpp
// Time Complexity : O(nk) + O(lk); n=number of words in dict; k=avg length of words; l=length of sentence
// Space Complexity : O(nk) - space of trie
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

// 1. Create a trie out of the dictionary and tokenize sentence into words
// 2. Search each word in trie, at any point if a char is not found return *, keep adding the char to prefix string
// 3. If input word is parsed or trie reaches char with isEnd=true, return prefix.
// 4. Keep forming result string by appending prefix/orig word (depending on if prefix exists) with a space. Remove final space separately

class TrieNode{
public:
    vector<TrieNode*> children;
    bool isEnd;
    
    TrieNode(){
        isEnd = false;
        children.resize(26);
        for(auto child: children)
            child=nullptr;
    }
    
    void createTrie(vector<string>& dict){
        for(string word: dict){
            TrieNode* root = this;
            for(char w: word){
                if(root->children[w-'a'] == nullptr)
                    root->children[w-'a'] = new TrieNode();
                root = root->children[w-'a'];
            }
            root->isEnd = true;
        }
    }
};

class Solution {
public:
    string findPrefix(TrieNode* root, string word){
        string prefix="";
        for(char w: word){
            if(root->children[w-'a'] == nullptr)
                return "*";
            prefix+=w;
            root=root->children[w-'a'];  
            if(root->isEnd == true)
                return prefix;
        }
        return prefix;
    }
    
    string replaceWords(vector<string>& dict, string sentence) {
        TrieNode* root = new TrieNode();
        root->createTrie(dict);
        stringstream sstr(sentence);
        string word, result;;
        while(getline(sstr, word, ' ')){
            string prefix = findPrefix(root, word);
            if(prefix=="*")
                result += word; 
            else
                result += prefix;
            result+=" ";
        }
        result.pop_back();
        return result;
    }
};

File: replace_word.java
//Time Complexity: O(n*k) + O(m*k)
   // n - words
    //m - words in sentence
    //k - max length of word
  //  Space Complexity: O(n*k)

  class Solution {
    class TrieNode {
        boolean isEnd;
        TrieNode children[];
        TrieNode() {
            children = new TrieNode[26];
        }
    }
    
    public void insert(TrieNode root, String word) {
        
        TrieNode current = root;
        for(int i =0; i<word.length();i++) {
            if(current.children[word.charAt(i)-'a'] == null) {
                current.children[word.charAt(i)-'a'] = new TrieNode();
            }
            current = current.children[word.charAt(i)-'a'];
        }
        current.isEnd = true;
    }
    public String replaceWords(List<String> dictionary, String sentence) {
        TrieNode root = new TrieNode();
        
        StringBuilder result = new StringBuilder();
        for(int i=0; i<dictionary.size();i++) {
            insert(root,dictionary.get(i));
        }
        String []senArray = sentence.split(" ");
        for(int i =0; i<senArray.length;i++) {
            TrieNode current = root;
            if(i != 0) {
                result.append(" ");
            }
            String word = senArray[i];
            StringBuilder sb = new StringBuilder();
          for(int j =0; j< word.length();j++) {
              if(current.children[word.charAt(j)-'a'] == null || current.isEnd) {
                  break;
              }
              current = current.children[word.charAt(j)-'a'];
              sb.append(word.charAt(j));

          }
          if(current.isEnd) {
              result.append(sb);
          } else {
              result.append(word);
          }
 }
 return result.toString();
    }
}